File: mParticle-Apple-SDK/Data Model/MPUpload.m
#import "MPUpload.h"
#import "MPIConstants.h"
#import "MPUploadSettings.h"
@import mParticle_Apple_SDK_Swift;

@interface MPUpload ()
@property (nonatomic, strong) MPUploadPRIVATE *implementation;
@end

@implementation MPUpload

- (instancetype)initWithSessionId:(NSNumber *)sessionId uploadDictionary:(NSDictionary *)uploadDictionary dataPlanId:(nullable NSString *)dataPlanId dataPlanVersion:(nullable NSNumber *)dataPlanVersion uploadSettings:(nonnull MPUploadSettings *)uploadSettings {
    NSData *uploadData = [MPUploadPRIVATE serializedUploadFromDictionary:uploadDictionary];
    if (!uploadData) {
        return nil;
    }

    NSDictionary *safeDictionary = [NSJSONSerialization JSONObjectWithData:uploadData options:0 error:nil];
    return [self initWithSessionId:sessionId
                          uploadId:0
                              UUID:safeDictionary[kMPMessageIdKey]
                        uploadData:uploadData
                         timestamp:[safeDictionary[kMPTimestampKey] doubleValue]
                        uploadType:MPUploadTypeMessage
                        dataPlanId:dataPlanId
                   dataPlanVersion:dataPlanVersion
                    uploadSettings:uploadSettings];
}

- (instancetype)initWithSessionId:(NSNumber *)sessionId uploadId:(int64_t)uploadId UUID:(NSString *)uuid uploadData:(NSData *)uploadData timestamp:(NSTimeInterval)timestamp uploadType:(MPUploadType)uploadType dataPlanId:(nullable NSString *)dataPlanId dataPlanVersion:(nullable NSNumber *)dataPlanVersion uploadSettings:(nonnull MPUploadSettings *)uploadSettings {
    self = [super init];
    if (self) {
        _implementation = [[MPUploadPRIVATE alloc] initWithSessionId:sessionId
                                                            uploadId:uploadId
                                                                UUID:uuid
                                                          uploadData:uploadData
                                                           timestamp:timestamp
                                                          uploadType:uploadType
                                                          dataPlanId:dataPlanId
                                                     dataPlanVersion:dataPlanVersion];
        _uuid = uuid;
        _uploadSettings = uploadSettings;
    }
    return self;
}

- (NSString *)description {
    return [NSString stringWithFormat:@"Upload\n Id: %lld\n UUID: %@\n Content: %@\n timestamp: %.0f\n Data Plan: %@ %@\n", self.uploadId, self.uuid, [self dictionaryRepresentation], self.timestamp, self.dataPlanId, self.dataPlanVersion];
}

- (BOOL)isEqual:(MPUpload *)object {
    if (MPIsNull(object) || ![object isKindOfClass:[MPUpload class]]) {
        return NO;
    }
    return [self.implementation isEqualToUpload:object.implementation];
}

- (NSUInteger)hash {
    return (NSUInteger)self.implementation.hash;
}

- (id)copyWithZone:(NSZone *)zone {
    MPUploadPRIVATE *copyImpl = [self.implementation copyUpload];
    MPUpload *copyObject = [[MPUpload alloc] initWithSessionId:copyImpl.sessionId
                                                      uploadId:copyImpl.uploadId
                                                          UUID:copyImpl.uuid
                                                    uploadData:copyImpl.uploadData
                                                     timestamp:copyImpl.timestamp
                                                    uploadType:(MPUploadType)copyImpl.uploadType
                                                    dataPlanId:copyImpl.dataPlanId
                                               dataPlanVersion:copyImpl.dataPlanVersion
                                                uploadSettings:[self.uploadSettings copy]];
    copyObject.containsOptOutMessage = copyImpl.containsOptOutMessage;
    return copyObject;
}

- (NSNumber *)sessionId { return self.implementation.sessionId; }
- (void)setSessionId:(NSNumber *)sessionId { self.implementation.sessionId = sessionId; }
- (int64_t)uploadId { return self.implementation.uploadId; }
- (void)setUploadId:(int64_t)uploadId { self.implementation.uploadId = uploadId; }
- (NSString *)uuid { return self.implementation.uuid; }
- (void)setUuid:(NSString *)uuid { self.implementation.uuid = uuid; _uuid = uuid; }
- (NSData *)uploadData { return self.implementation.uploadData; }
- (void)setUploadData:(NSData *)uploadData { self.implementation.uploadData = uploadData; }
- (NSTimeInterval)timestamp { return self.implementation.timestamp; }
- (void)setTimestamp:(NSTimeInterval)timestamp { self.implementation.timestamp = timestamp; }
- (MPUploadType)uploadType { return (MPUploadType)self.implementation.uploadType; }
- (void)setUploadType:(MPUploadType)uploadType { self.implementation.uploadType = uploadType; }
- (BOOL)containsOptOutMessage { return self.implementation.containsOptOutMessage; }
- (void)setContainsOptOutMessage:(BOOL)containsOptOutMessage { self.implementation.containsOptOutMessage = containsOptOutMessage; }
- (NSString *)dataPlanId { return self.implementation.dataPlanId; }
- (void)setDataPlanId:(NSString *)dataPlanId { self.implementation.dataPlanId = dataPlanId; }
- (NSNumber *)dataPlanVersion { return self.implementation.dataPlanVersion; }
- (void)setDataPlanVersion:(NSNumber *)dataPlanVersion { self.implementation.dataPlanVersion = dataPlanVersion; }

- (NSDictionary *)dictionaryRepresentation {
    return [self.implementation dictionaryRepresentation];
}

- (NSString *)serializedString {
    return [self.implementation serializedString];
}

@end

File: mParticle-Apple-SDK-Swift/Sources/DataModel/MPUpload.swift
import Foundation

@objc(MPUpload)
public final class MPUploadPRIVATE: NSObject, NSCopying {
    @objc public var sessionId: NSNumber?
    @objc public var uploadId: Int64
    @objc public var uuid: String?
    @objc public var uploadData: Data
    @objc public var timestamp: TimeInterval
    @objc public var uploadType: UInt
    @objc public var containsOptOutMessage = false
    @objc public var dataPlanId: String?
    @objc public var dataPlanVersion: NSNumber?
    @objc public var uploadSettings: NSObject

    @objc(initWithSessionId:uploadId:UUID:uploadData:timestamp:uploadType:dataPlanId:dataPlanVersion:uploadSettings:)
    public init(
        sessionId: NSNumber?,
        uploadId: Int64,
        uuid: String?,
        uploadData: Data,
        timestamp: TimeInterval,
        uploadType: UInt,
        dataPlanId: String?,
        dataPlanVersion: NSNumber?,
        uploadSettings: NSObject
    ) {
        self.sessionId = sessionId
        self.uploadId = uploadId
        self.uuid = uuid
        self.uploadData = uploadData
        self.timestamp = timestamp
        self.uploadType = uploadType
        self.dataPlanId = dataPlanId
        self.dataPlanVersion = dataPlanVersion
        self.uploadSettings = uploadSettings
        super.init()
    }

    @objc(initWithSessionId:uploadDictionary:dataPlanId:dataPlanVersion:uploadSettings:)
    public convenience init?(
        sessionId: NSNumber?,
        uploadDictionary: NSDictionary,
        dataPlanId: String?,
        dataPlanVersion: NSNumber?,
        uploadSettings: NSObject
    ) {
        guard let uploadData = Self.serializedUpload(from: uploadDictionary),
              let safeDictionary = try? JSONSerialization.jsonObject(with: uploadData, options: []) as? NSDictionary
        else {
            return nil
        }

        self.init(
            sessionId: sessionId,
            uploadId: 0,
            uuid: safeDictionary["id"] as? String,
            uploadData: uploadData,
            timestamp: (safeDictionary["ct"] as? NSNumber)?.doubleValue ?? 0,
            uploadType: 0,
            dataPlanId: dataPlanId,
            dataPlanVersion: dataPlanVersion,
            uploadSettings: uploadSettings
        )
    }

    @objc(serializedUploadFromDictionary:)
    public static func serializedUpload(from dictionary: NSDictionary) -> Data? {
        do {
            guard let safeDictionary = MPJSONCopyPRIVATE.deepCopyJSONObject(dictionary) as? NSDictionary,
                  JSONSerialization.isValidJSONObject(safeDictionary)
            else {
                return nil
            }
            return try JSONSerialization.data(withJSONObject: safeDictionary, options: [])
        } catch {
            NSLog("mParticle -> Exception serializing upload dictionary: %@", error.localizedDescription)
            return nil
        }
    }

    @objc public func dictionaryRepresentation() -> NSDictionary? {
        (try? JSONSerialization.jsonObject(with: uploadData, options: [])) as? NSDictionary
    }

    @objc public func serializedString() -> String? {
        String(data: uploadData, encoding: .utf8)
    }

    @objc public func copy(with _: NSZone? = nil) -> Any {
        let settings = (uploadSettings as? NSCopying)?.copy(with: nil) as? NSObject ?? uploadSettings
        let copy = MPUploadPRIVATE(
            sessionId: sessionId,
            uploadId: uploadId,
            uuid: uuid,
            uploadData: uploadData,
            timestamp: timestamp,
            uploadType: uploadType,
            dataPlanId: dataPlanId,
            dataPlanVersion: dataPlanVersion,
            uploadSettings: settings
        )
        copy.containsOptOutMessage = containsOptOutMessage
        return copy
    }

    override public func isEqual(_ object: Any?) -> Bool {
        guard let other = object as? MPUploadPRIVATE else { return false }
        return isEqual(toUpload: other)
    }

    @objc(isEqualToUpload:)
    public func isEqual(toUpload other: MPUploadPRIVATE) -> Bool {
        sessionId == other.sessionId
            && uploadId == other.uploadId
            && timestamp == other.timestamp
            && dataPlanId == other.dataPlanId
            && dataPlanVersion == other.dataPlanVersion
    }

    override public var hash: Int {
        (sessionId?.hashValue ?? 0)
            ^ (dataPlanId?.hashValue ?? 0)
            ^ (dataPlanVersion?.hashValue ?? 0)
            ^ Int(truncatingIfNeeded: uploadId)
            ^ Int(timestamp)
    }

    override public var description: String {
        "Upload\n Id: \(uploadId)\n UUID: \(uuid ?? "(null)")\n Content: "
            + "\(dictionaryRepresentation()?.description ?? "(null)")\n timestamp: \(String(format: "%.0f", timestamp))"
            + "\n Data Plan: \(dataPlanId ?? "(null)") \(dataPlanVersion?.description ?? "(null)")\n"
    }
}
